pkg/controller/endpoint/endpoints_controller.go

```diff
@@ -184,18 +188,36 @@ func (e *EndpointController) Run(workers int, stopCh <-chan struct{}) {
 
 func (e *EndpointController) getPodServiceMemberships(pod *v1.Pod) (sets.String, error) {
 	set := sets.String{}
-	services, err := e.serviceLister.GetPodServices(pod)
+	allServices, err := e.serviceLister.Services(pod.Namespace).List(labels.Everything())
 	if err != nil {
 		// don't log this error because this function makes pointless
 		// errors when no services match.
 		return set, nil
 	}
-	for i := range services {
-		key, err := controller.KeyFunc(services[i])
-		if err != nil {
-			return nil, err
+
+	for _, service := range allServices {
+		var selector labels.Selector
+		tmp, ok := e.serviceSelectorCache.Load(service)
+		if ok {
+			// retrieve selector from cache
+			selector = tmp.(labels.Selector)
+		} else {
+			// only create a selector if we don't have it cached already
+			if service.Spec.Selector == nil {
+				// services with nil selectors match nothing, not everything.
+				continue
+			}
+			selector = labels.Set(service.Spec.Selector).AsSelectorPreValidated()
+			e.serviceSelectorCache.Store(service, selector)
+		}
+
```

Comment:
In this solution there is no garbage collection from the map, so it will be infinitely growing. That's not something we can accept.

Reply:
Fixed, handled during `syncService`